surveys/promise-rejections/survey.md

```diff
@@ -0,0 +1,235 @@
+
+## What is a Promise rejection?
+
+A Promise rejection indicates that something went wrong while executing a Promise or an async function. Rejections can occur in several situations: throwing inside an async function or a `Promise` executor/then/catch/finally callback, when calling the `reject` callback of an `executor`, or when calling `Promise.reject`.
+
+```js
+Promise.reject(new Error())  // This will result in a rejection
+
+new Promise((fulfill, reject) => {
+  reject(new Error());  // This will result in a rejection
+});
+
+new Promise(() => {
+  throw new Error();  // This will result in a rejection
+})
+
+Promise.resolve().then(() => {
+  throw new Error();  // This will result in a rejection
+})
+
+Promise.reject().then(() = {}, () => {
+  throw new Error();  // This will result in a rejection
+}
+
+Promise.reject().catch(() => {
+  throw new Error();  // This will result in a rejection
+})
+
+Promise.resolve().finally(() => {
+  throw new Error();  // This will result in a rejection
+});
+
+async function () {
+  throw new Error();  // This will result in a rejection
+}
+
+async function () {
+  if (a === undefined)
+    a();  // This will result in a rejection
+}
+```
+
+Adding the `async` keyword to a function will turn any exceptions thrown inside that function (or any throw propagated from other functions called within it) into a rejection. The same happens when refactoring callback based code that throws into async functions / Promises. Below is an example of a callback based code refactored to Promises where exceptions become rejections:
+
+## What is an unhandled rejection?
+
+There are two ways to handle rejections: by attaching a `.catch` handler to it,
+or by awaiting on the promise within a try/catch block. In both cases, the
+handling of the rejection (the execution of the callback passed to `.catch`, or
+the execution of the `catch {}` block) will happen in a future turn of the
+event loop.
+
+Promises are designed so that attaching handlers or awaiting can be done at any
+point in time, from when the Promise was created (possibly while it's still
+pending), to right before the program finishes execution.
+
+A rejection is considered unhandled from the point it happens until the point
+where a handler is attached to the Promise or the Promise is awaited within a
+`catch {}` block. Below are a few examples of handled and unhandled rejections.
+
+```js
+async function foo() {
+  throw new Error();
+}
+
+foo()  // 1. Unhandled at this point
+  .catch(() => console.error("an error occured"));  // 2. Now it's handled
+
+try {
+  await foo();
+} catch(e) {  // 3. Handled
+  console.error("an error occured")
+}
+
+foo(); // 4. Unhandled, but execution continues
+
+const rejected = foo(); // 5. Unhandled on current event loop turn
+
+// 4. ... but handled in a future turn of the event loop when the setTimeout
```

Comment:
The numbering here is inconsistent... there are two points numbered 4, and the "example 4" text below it is confusing.

Reply:
Oops, I messed up the numbering here. I'll fix it.

> and the "example 4" text below it is confusing

Any suggestions on how to make it more approachable?

Reply:
I just meant that you can't say "example 4" when there are two examples 4 ;-)

Reply:
Oh, I see. So the idea is that the "second" example 4 is a continuation of the first (it's still the same example, but it shows how a catch can be attached at a later point in time). I'll see if there's a way to rephrase and reorganize so it keeps the message but in a clear way (or, if you have any suggestions, I'm open to hear :) )